src/node/txdownload_impl.cpp

```diff
@@ -0,0 +1,415 @@
+    //   - OVERLOADED_PEER_TX_DELAY for announcements from peers which have at least
+    //     MAX_PEER_TX_REQUEST_IN_FLIGHT requests in flight (and don't have NetPermissionFlags::Relay).
+    auto delay{0us};
+    if (!info.m_preferred) delay += NONPREF_PEER_TX_DELAY;
+    if (!gtxid.IsWtxid() && m_num_wtxid_peers > 0) delay += TXID_RELAY_DELAY;
+    const bool overloaded = !info.m_relay_permissions && m_txrequest.CountInFlight(peer) >= MAX_PEER_TX_REQUEST_IN_FLIGHT;
+    if (overloaded) delay += OVERLOADED_PEER_TX_DELAY;
+
+    m_txrequest.ReceivedInv(peer, gtxid, info.m_preferred, now + delay);
+}
+
+void TxDownloadImpl::ReceivedTxInv(NodeId peer, const GenTxid& gtxid, std::chrono::microseconds now)
+    EXCLUSIVE_LOCKS_REQUIRED(!m_tx_download_mutex)
+{
+    LOCK(m_tx_download_mutex);
+    AddTxAnnouncement(peer, gtxid, now);
+}
+
+std::vector<GenTxid> TxDownloadImpl::GetRequestsToSend(NodeId nodeid, std::chrono::microseconds current_time)
+    EXCLUSIVE_LOCKS_REQUIRED(!m_tx_download_mutex)
+{
+    LOCK(m_tx_download_mutex);
+    // First process orphan resolution so that the tx requests can be sent asap
+    std::vector<std::pair<NodeId, GenTxid>> expired_orphan_resolution;
+    const auto orphans_ready = m_orphan_resolution_tracker.GetRequestable(nodeid, current_time, &expired_orphan_resolution);
+    // Expire orphan resolution attempts
+    for (const auto& [nodeid, orphan_gtxid] : expired_orphan_resolution) {
+        LogPrintf("timeout of in-flight orphan resolution %s for peer=%d\n", orphan_gtxid.GetHash().ToString(), nodeid);
+        // All txhashes in m_orphan_resolution_tracker are wtxids.
+        Assume(orphan_gtxid.IsWtxid());
+        m_orphanage.EraseOrphanOfPeer(orphan_gtxid.GetHash(), nodeid);
+    }
+    for (const auto& orphan_gtxid : orphans_ready) {
+        Assume(orphan_gtxid.IsWtxid());
+        const auto parent_txids{m_orphanage.GetParentTxids(orphan_gtxid.GetHash())};
+        if (parent_txids.has_value()) {
+            if (!Assume(m_peer_info.count(nodeid) > 0)) continue;
+            const auto& info = m_peer_info.at(nodeid).m_connection_info;
+            for (const auto& txid : *parent_txids) {
+                // Schedule with no delay. It should be requested immediately
+                // unless there is already a request out for this transaction.
+                m_txrequest.ReceivedInv(nodeid, GenTxid::Txid(txid), info.m_preferred, current_time);
+                LogPrint(BCLog::TXPACKAGES, "scheduled parent request %s from peer=%d for orphan %s\n",
+                         txid.ToString(), nodeid, orphan_gtxid.GetHash().ToString());
+            }
+            m_orphan_resolution_tracker.RequestedTx(nodeid, orphan_gtxid.GetHash(),
+                                                    current_time + ORPHAN_ANCESTOR_GETDATA_INTERVAL);
+        } else {
+            LogPrint(BCLog::TXPACKAGES, "couldn't find parent txids to resolve orphan %s with peer=%d\n",
```

Comment:
under what circumstances is this expected? Eviction of the orphan in the orphanage, but outstanding resolution request? A comment to the expected conditions would be helpful for documentation.

Reply:
ok, it's happening because of: https://github.com/bitcoin/bitcoin/pull/28031/files#r1309300862